buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java

```diff
@@ -907,39 +1051,58 @@ private boolean allocateWithoutLock(int size, int maxCapacity, AdaptiveByteBuf b
+                        curr.readInitInto(buf, size, -1, maxCapacity);
                         return true;
-                    } finally {
-                        curr.releaseFromMagazine();
                     }
-                }
-
-                // Check if we either retain the chunk in the nextInLine cache or releasing it.
-                if (remainingCapacity < RETIRE_CAPACITY) {
+                    // Chunk is full. Release and fall through to get a new one.
+                    current = null;
                     curr.releaseFromMagazine();
                 } else {
-                    // See if it makes sense to transfer the Chunk to the nextInLine cache for later usage.
-                    // This method will release curr if this is not the case
-                    transferToNextInLineOrRelease(curr);
+                    int startingCapacity = chunkController.computeBufferCapacity(size, maxCapacity, reallocate);
+                    int remainingCapacity = curr.remainingCapacity();
+                    if (remainingCapacity > startingCapacity) {
+                        curr.readInitInto(buf, size, startingCapacity, maxCapacity);
+                        // We still have some bytes left that we can use for the next allocation, just early return.
+                        return true;
+                    }
+
+                    // At this point we know that this will be the last time current will be used, so directly set it to
+                    // null and release it once we are done.
+                    current = null;
+                    if (remainingCapacity >= size) {
+                        try {
+                            curr.readInitInto(buf, size, remainingCapacity, maxCapacity);
+                            return true;
+                        } finally {
+                            curr.releaseFromMagazine();
+                        }
+                    }
+
+                    // Check if we either retain the chunk in the nextInLine cache or releasing it.
+                    if (remainingCapacity < RETIRE_CAPACITY) {
+                        curr.releaseFromMagazine();
+                    } else {
+                        // See if it makes sense to transfer the Chunk to the nextInLine cache for later usage.
+                        // This method will release curr if this is not the case
+                        transferToNextInLineOrRelease(curr);
+                    }
                 }
             }
 
+            // current is null or was exhausted.
+            int startingCapacity = chunkController.computeBufferCapacity(size, maxCapacity, reallocate);
+            return allocateAndRefillCurrentChunk(size, maxCapacity, buf, startingCapacity);
+        }
+
+        private boolean allocateAndRefillCurrentChunk(int size, int maxCapacity, AdaptiveByteBuf buf,
+                                                      int startingCapacity) {
+            Chunk curr;
```

Comment:
I expect a volatile `nextInLine` load first to gate `curr = NEXT_IN_LINE.getAndSet(this, null);`